src/ray/common/scheduling/scheduling_alg.cc

```diff
@@ -0,0 +1,204 @@
+#include "scheduling_alg.h"
+#include <iostream>
+#include <stdlib.h>
+#include <chrono>
+
+using namespace std;
+
+ClusterResources::ClusterResources(int64_t local_node_id, const NodeResources &node_resources) {
+  local_node_id_ = local_node_id;
+  add(local_node_id, node_resources);
+}
+
+void ClusterResources::add(int64_t node_id, const NodeResources &node_resources) {
+
+  auto it = nodes_.find(node_id);
+  if (it == nodes_.end()) {
+    // This node is new. Add it to the list.
+    nodes_.emplace(node_id, node_resources);
+  } else {
+    // This node exists. Update it's resources.
+    NodeResources &nr = it->second;
+
+    // First, update its predefined resources.
+    for (int i = 0; i < NUM_PREDIFINED_RESOURCES; i++) {
+      nr.capacities[i].total = node_resources.capacities[i].total;
+      nr.capacities[i].available = node_resources.capacities[i].available;
+    }
+
+    // Then, delete existing custom resources and replace them with the new ones.
+    nr.custom_resources.clear();
+
+    auto mcr = node_resources.custom_resources;
+    for (auto it = mcr.begin(); it != mcr.end(); ++it) {
+      nr.custom_resources.insert(*it);
+    }
+  }
+}
+
+bool ClusterResources::remove(int64_t node_id) {
+  auto it = nodes_.find(node_id);
+  if (it == nodes_.end()) {
+    /// Node not found.
+    return false;
+  } else {
+    /// Remobe node.
+    it->second.custom_resources.clear();
+    nodes_.erase(it);
+    return true;
+  }
+}
+
+int64_t ClusterResources::isSchedulable(const TaskReq &task_req, const NodeResources &nr) {
+  int violations = 0;
+
+  /// First, check predefined resources.
+  for (int i = 0; i < NUM_PREDIFINED_RESOURCES; i++) {
+    if (task_req.predefined_resources[i].demand > nr.capacities[i].available) {
+      if (task_req.predefined_resources[i].soft) {
+        // A soft constraint has been violated.
+        violations++;
+      } else {
+        // A hard constraint has been violated.
+        return -1;
+      }
+    }
+  }
+
+  for (int i = 0; i < task_req.custom_resources.size(); i++) {
+    auto it = nr.custom_resources.find(task_req.custom_resources[i].id);
+
+    if (it == nr.custom_resources.end()) {
+      /// Requested resource doesn't exist at this node.
+      if (task_req.custom_resources[i].req.soft) {
+        violations++;
+      } else {
+        return -1;
+      }
+    } else {
+      if (task_req.custom_resources[i].req.demand > it->second.available) {
+        /// Resource constraint is violated.
+        if (task_req.custom_resources[i].req.soft) {
+          violations++;
+        } else {
+          return -1;
+        }
+      }
+    }
+  }
+  return violations;
+}
+
+int64_t ClusterResources::getSchedulableNode(const TaskReq &task_req, int64_t *violations) {
```

Comment:
This function is pretty complicated, I think it would be much nicer if you could break it into a few separate, smaller functions so the high level logic is clearer from reading it.